src/backends/perf_pt/mod.rs

```diff
@@ -0,0 +1,221 @@
+// (b) any piece of software and/or hardware listed in the lrgrwrks.txt file
+// if one is included with the Software (each a "Larger Work" to which the Software
+// is contributed by such licensors),
+//
+// without restriction, including without limitation the rights to copy, create
+// derivative works of, display, perform, and distribute the Software and make,
+// use, sell, offer for sale, import, export, have made, and have sold the
+// Software and the Larger Work(s), and to sublicense the foregoing rights on
+// either these or other terms.
+//
+// This license is subject to the following condition: The above copyright
+// notice and either this complete permission notice or at a minimum a reference
+// to the UPL must be included in all copies or substantial portions of the
+// Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+use libc::{pid_t, c_char, c_void, getpid, size_t, c_int, geteuid};
+use std::ffi::CString;
+use std::path::PathBuf;
+use errors::TraceMeError;
+use std::fs::File;
+use std::io::Read;
+use Tracer;
+
+// The sysfs path used to set perf permissions.
+const PERF_PERMS_PATH: &'static str = "/proc/sys/kernel/perf_event_paranoid";
+
+// FFI prototypes.
+extern "C" {
+    fn perf_pt_start_tracer(conf: *const PerfPTConf) -> *const c_void;
+    fn perf_pt_stop_tracer(tr_ctx: *const c_void) -> c_int;
+}
+
+// Struct used to communicate a tracing configuration to the C code. Must
+// stay in sync with the C code.
+#[repr(C)]
+struct PerfPTConf {
+    target_pid: pid_t,
+    trace_filename: *const c_char,
+    map_filename: *const c_char,
+    data_bufsize: size_t,
+    aux_bufsize: size_t,
+}
+
+/// A tracer that uses the Linux Perf interface to Intel Processor Trace.
+#[derive(Default)]
+pub struct PerfPTTracer {
+    /// Filename to store the trace to.
+    trace_filename: String,
+    /// PID to trace.
+    target_pid: pid_t,
+    /// Data buffer size, in pages. Must be a power of 2.
+    data_bufsize: size_t,
+    /// Aux buffer size, in pages. Must be a power of 2.
+    aux_bufsize: size_t,
+    /// Opaque C pointer representing the tracer context.
+    tracer_ctx: Option<*const c_void>,
+}
+
+impl PerfPTTracer {
+    /// Create a new tracer.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use traceme::backends::PerfPTTracer;
+    ///
+    /// PerfPTTracer::new().trace_filename("mytrace.ptt").data_bufsize(1024).target_pid(666);
+    /// ```
+    pub fn new() -> Self {
+        Self {
+            trace_filename: String::from("traceme.ptt"),
```

Comment:
Having a fixed filename feels weird to me. Should it be configurable?

Reply:
It is configurable, yes:
https://github.com/softdevteam/traceme/blob/678ff424831a866e760fd4b242278e24e8318222/src/backends/perf_pt/mod.rs#L101-L109

However, don't worry too much about the trace filenames. I'm going to tear this out when I work on #8. We don't want to trace to disk, as it will be too slow.

[I might add the ability to dump traces to disk as a debug facility only (and the i'd derive the trace filename from the pid and the address of the `Location`)]

Reply:
OK.